chips/sam4l/src/eic.rs

```diff
@@ -0,0 +1,321 @@
+
+        // Writing a zero to this bit will set the input value to INTn to zero, if test mode is enabled. 
+        // Writing a one to this bit will set the input value to INTn to one, if test mode is enabled.
+        INT OFFSET(0) NUMBITS(31) []
+    ]
+];
+
+// Page 59 of SAM4L data sheet
+const EIC_BASE: StaticRef<EicRegisters> =
+    unsafe { StaticRef::new(0x400F1000 as *const EicRegisters) };
+
+pub struct Eic<'a> {
+    registers: StaticRef<EicRegisters>,
+    enabled: Cell<bool>,
+    callbacks: [OptionalCell<&'a dyn hil::eic::Client>; 9],
+}
+
+impl<'a> hil::eic::ExternalInterruptController for Eic<'a> {
+    type Line = Line;
+
+    fn line_enable(
+        &self,
+        line: &Self::Line,
+        int_mode: hil::eic::InterruptMode,
+        filter_mode: hil::eic::FilterMode,
+        syn_mode: hil::eic::SynchronizationMode,
+    ) {
+        if !self.is_enabled() {
+            self.enable();
+        }
+
+        let interrupt_line = *line as u32;
+        let regs: &EicRegisters = &*self.registers;
+
+        regs.en.write(Interrupt::INT.val(interrupt_line));
+
+        self.line_configure(interrupt_line, int_mode, filter_mode, syn_mode);
+        self.line_enable_interrupt(interrupt_line);
+    }
+
+    fn line_disable(&self, line: &Self::Line) {
+        if !self.is_enabled() {
+            return;
+        }
+
+        let interrupt_line: u32 = 1 << *line as u32;
+        let regs: &EicRegisters = &*self.registers;
+        regs.dis.write(Interrupt::INT.val(interrupt_line));
+        self.line_disable_interrupt(interrupt_line);
+    }
+}
+
+impl<'a> Eic<'a> {
+    pub fn line_configure(
+        &self,
+        interrupt_line: u32,
+        int_mode: hil::eic::InterruptMode,
+        filter_mode: hil::eic::FilterMode,
+        syn_mode: hil::eic::SynchronizationMode,
+    ) {
+        let mode_bits = match int_mode {
+            hil::eic::InterruptMode::RisingEdge => 0b00,
+            hil::eic::InterruptMode::FallingEdge => 0b01,
+            hil::eic::InterruptMode::HighLevel => 0b10,
+            hil::eic::InterruptMode::LowLevel => 0b11,
+        };
+
+        self.set_interrupt_mode(mode_bits, interrupt_line);
+
+        match filter_mode {
+            hil::eic::FilterMode::FilterEnable => self.line_enable_filter(interrupt_line),
+            hil::eic::FilterMode::FilterDisable => self.line_disable_filter(interrupt_line),
+        }
+
+        match syn_mode {
+            hil::eic::SynchronizationMode::Synchronous => self.line_disable_asyn(interrupt_line),
+            hil::eic::SynchronizationMode::Asynchronous => self.line_enable_asyn(interrupt_line),
+        }
+    }
+
+    pub fn enable(&self) {
+        pm::enable_clock(Clock::PBD(PBDClock::EIC));
+        self.enabled.set(true);
+    }
+
+    pub fn disable(&self) {
+        pm::disable_clock(Clock::PBD(PBDClock::EIC));
+        self.enabled.set(false);
+    }
+
+    pub fn set_interrupt_mode(&self, mode_bits: u8, interrupt_line: u32) {
```

Comment:
Should this be `pub`? It looks like it's only called as a helper method by `line_configure`. The reason I ask is that here the function is taking an opaque set of `mode_bits` where `line_configure` takes a nice `InterruptMode` type. This is interface is fine if this is internal (non-`pub`), but if it's meant to be external facing, it should probably use the better type in the interface

Reply:
This should be private. Thanks for catching that! I've fixed others as well. For those functions that return a bool type, I used them mostly for testing purposes and made them pub but the users wouldn't use these functions in any scenario; is it better to remove these functions or leave them be?

Reply:
Hmm.., this is trickier -- we don't have the greatest modular testing story in Tock right now. There are a few places that use `#[cfg(test)]`, but I'll confess to ignorance of the details. I'm inclined not to leave test code lying around, especially with pub visibility, but I'll defer to others on the best way to handle this.

Reply:
I think `#[cfg(test)]` doesn't work because it pulls in `std` however https://doc.rust-lang.org/unstable-book/language-features/custom-test-frameworks.html looks like an option or define our own feature flag for test builds